src/anndata/_core/anndata.py

```diff
@@ -662,6 +662,10 @@ def X(self, value: np.ndarray | sparse.spmatrix | SpArray | None):
                         else:
                             memory_class = sparse.coo_matrix
                         value = memory_class(value)
+                    elif sparse.issparse(value) and isinstance(
+                        self._adata_ref._X, np.ndarray
+                    ):
+                        value = value.toarray()
```

Comment:
We definitely could warn here, but on the other hand, we don't warn in the above case when we sparsify from dense (which entails a >2x memory increase)